cypress-io/cypress
# fix: Add missing header and use correct endpoint host

- Closes https://github.com/cypress-io/cypress/issues/23977

### User facing changelog


n/a

### Additional details


Events being recorded were not being sent to the correct host and were missing a required header.  This change adds the missing `Content-Type` header, adjusts the logic for selecting the correct host for the endpoint, and documents findings regarding how the CYPRESS_INTERNAL_ENV variable gets set.

### Steps to test


The original PR has full testing instructions [here](You can use the testing instructions in the original PR that introduced this feature here: https://github.com/cypress-io/cypress/pull/23691)

For this fix, you can reduce the testing to the following:

* Go to the `packages/app` directory. 
* Start the app by running `DEBUG=cypress:data-context:actions:EventCollectorActions yarn dev`
* Remove any `banners` entries in the `state.json` for this project.  You find this file by going to "View App Data"  in the "Developer Tools" menu
* Confirm that the timestamp for "firstOpened" in the `state.json` file is older than 4 days. 
* Make sure you are logged out of the dashboard
* Open either the e2e or component project. 
* Confirm on the command line that you see similar to the following:
```
[cy:open:dev:33905]:   cypress:data-context:actions:EventCollectorActions Using staging environment for Event Collection +0ms
[cy:open:dev:33905]:   cypress:data-context:actions:EventCollectorActions Recorded event: { campaign: 'Log In', messageId: 'GtvrFq6I1pUdGsBijOSJ0', medium: 'Specs Login Banner', cohort: 'B' } +222ms
```
* Look to make sure that the 'staging' environment is being used.
* Quit the Cypress app
* Edit the 'state.json' file again to remove any `banners` entries.
* Start the app with `DEBUG=cypress:data-context:actions:EventCollectorActions CYPRESS_INTERNAL_ENV=production yarn dev`
* Run the same test again by opening either the e2e or component tests for the project
* Confirm you see similar to the following on the command line:
```
[cy:open:dev:33905]:   cypress:data-context:actions:EventCollectorActions Using production environment for Event Collection +0ms
[cy:open:dev:33905]:   cypress:data-context:actions:EventCollectorActions Recorded event: { campaign: 'Log In', messageId: 'GtvrFq6I1pUdGsBijOSJ0', medium: 'Specs Login Banner', cohort: 'B' } +222ms
```
* Look to make sure the 'production' environment is being used

To test the built binary:
* Go to the root directory of the project and run `DEBUG=cypress:binary yarn binary-build --skipSigning`
* It should take about 2 minutes and you should see at the end a green checkmark and it says `build completed`
* Scroll up above the smoke tests to look for a green checkmark that says "using node --version works".  Just below that should be a line that says "executable path" with a path into a temporary folder pointing at the built executable
* Run the app using the same debug as above with this path.  Something like:
```
DEBUG=cypress:data-context:actions:EventCollectorActions /var/folders/lv/40hdk3vx1m96w83cc8t820800000gq/T/cypress-build/darwin/build/mac-arm64/Cypress.app/Contents/MacOS/Cypress
```
* This will open in global mode.  
* Pick a project or add one if you do not have one.  
* Confirm the same as before that the `state.json` does not have any `banners` entries and that you are not logged in
* Open the project in either e2e or component testing.
* Confirm on the command line that you see the same debug messages for the previous tests and that it says `Using production environment for Event Collection`

### How has the user experience changed?


n/a

### PR Tasks


- [ n/a] Have tests been added/updated?
- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
